app/grandchallenge/algorithms/forms.py

```diff
@@ -365,21 +365,77 @@ class Meta:
             "workstation_config": "Viewer Configuration",
         }
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(self, *args, user, **kwargs):
+        super().__init__(*args, user=user, **kwargs)
+        self._user = user
 
         self.fields["contact_email"].required = True
         self.fields["display_editors"].required = True
 
-        self.fields["job_requires_gpu_type"].choices = [
-            (choice.value, choice.label)
-            for choice in [GPUTypeChoices.NO_GPU, GPUTypeChoices.T4]
-        ]
+        self.fields["job_requires_gpu_type"].choices = (
+            self.get_selectable_gpu_type_choices()
+        )
         self.fields["job_requires_memory_gb"].validators = [
             MinValueValidator(settings.ALGORITHMS_MIN_MEMORY_GB),
-            MaxValueValidator(settings.ALGORITHMS_MAX_MEMORY_GB),
+            MaxValueValidator(self.get_maximum_settable_memory_gb()),
         ]
 
+    @cached_property
+    def relevant_phases(self):
```

Comment:
This seems relatively familiar, was it duplicated from somewhere?

Reply:
Yes, I copied it from the AlgorithmsForPhaseMixin, [this part (L396-436)](https://github.com/comic/grand-challenge.org/pull/3714/files/f8560de9d07fba4a3bdd505a37d298cd9a890cf1#diff-800ed244caee41acb2694b83ed490082bcfe999937aa38f5e3cf1f4fdf0ef8ceL396-L436)

Reply:
Is there a way to reuse this rather than duplicating it?

Reply:
Sorry, I didn't mean it is an exact copy. I just copied the way the queryset uses annotations and then filters them. The code there is actually doing the opposite (filtering for algorithms based on a phase, instead of filtering for phases based on an algorithm). I don't see a way of generalising this.

Reply:
Okay that's fine!